Ftr/dynamic annotation classification

bwbohl opened the pull request:
## Description, Context and related Issue
Add support for dynamic mei:taxonomy usage in mei annot

<!--- This project only accepts pull requests related to open issues. Please link to the issue here: -->
Refs https://github.com/Edirom/Edirom-Online-Frontend/issues/57

## How Has This Been Tested?
* Klarinettenquintett dataset
* BAZ-GA dataset

## Types of changes
- New feature (non-breaking change which adds functionality)
- Documentation Update
- Improvement
- Refactoring

## Overview
- I have updated the inline documentation accordingly.
- I have performed a self-review of my code, according to the [style guide](https://github.com/Edirom/Edirom-Online/blob/develop/STYLE-GUIDE.md)
- I have read the [CONTRIBUTING](https://github.com/Edirom/Edirom-Online-Backend/blob/develop/CONTRIBUTING.md) document.


## ToDo:s / things to check

- [x] **(1) Annotation view display**
  When I open an annotation (click on annotation marker) in local Klarinettenquintett, the whole annotation list opens and I can't switch to single annotation view. In comparison, [online Klarinettenquintett](https://klarinettenquintett.weber-gesamtausgabe.de/index.html) displays single annotation view by default.

- [x] **(2) Annotation view meta info box**
  Titles of annotation taxonomy classes are shown in single annotation view
 
- [x] **(3) Annotation menu**
  Newly added annotation category does not show up in menu:

- [x] **(4) Annotation list**
  Newly added annotation category shows up in complete annotation list with translation

- [x] **(5) Translation injection**
And one more side-note: Is it possible to inject translations from edition xar into backend? That would be really good for this feature.

- [x] **(6) PageBasedView support**
- [x] **(7) MeasureBasedView support**
- [x] **(8) Test with legacy OpenSeadragonViewer**
- [x] **(9) Test with legacy ImageViewer**
- [x] **(10) maintain old classification model support** (category / priority) --> test with old data!

bwbohl committed:
AnnotationView: mark TODO

bwbohl committed:
AnnotationView: mark TODO

bwbohl committed:
AnnotationView: use calculated store fields instead of model

bwbohl committed:
AnnotationView: use calculated store fields instead of model

bwbohl committed:
AnnotationView: add method for calculating columns

bwbohl committed:
AnnotationView: add method for calculating columns

bwbohl committed:
AnnotationView: remove predefined me.list

bwbohl committed:
AnnotationView: remove predefined me.list

bwbohl committed:
AnnotationView: add method for creating list and store

bwbohl committed:
AnnotationView: add method for creating list and store

bwbohl committed:
AnnotationView: add method for loading Annotations and creating components

bwbohl committed:
AnnotationView: add method for loading Annotations and creating components

bwbohl committed:
AnnotationView: create method for initializing independent components

bwbohl committed:
AnnotationView: create method for initializing independent components

bwbohl committed:
AnnotationView: add event and property annotationsLoaded

bwbohl committed:
AnnotationView: add event and property annotationsLoaded

bwbohl committed:
AnnotationView: call loadAnnotationsAndCreateComponents afterrender

bwbohl committed:
AnnotationView: call loadAnnotationsAndCreateComponents afterrender

bwbohl committed:
AnnotationView: remove forgotten event listener

bwbohl committed:
AnnotationView: remove forgotten event listener

bwbohl committed:
AnnotationsView: load list store only if defined

bwbohl committed:
AnnotationsView: load list store only if defined

bwbohl committed:
AnnotationView: showList add check if me.list available

bwbohl committed:
AnnotationView: showList add check if me.list available

bwbohl committed:
Delete Annotation.js

bwbohl committed:
Delete Annotation.js

bwbohl committed:
AnnotationView: rework column definitions to hide empty columns

bwbohl committed:
AnnotationView: rework column definitions to hide empty columns

bwbohl committed:
AnnotationView: getLangString for column headers

bwbohl committed:
AnnotationView: getLangString for column headers

bwbohl committed:
AnnotationView: add default column definition for `id`

in order to achieve the following:
* use existing lang key that does not match the calculated key form
* set to hidden by default

bwbohl committed:
AnnotationView: add default column definition for `id`

in order to achieve the following:
* use existing lang key that does not match the calculated key form
* set to hidden by default

bwbohl committed:
AnnotationView:  readd default column definition for title

to achieve the following:
* use existing lang key that does not match the calculated key
* increase flex size

bwbohl committed:
AnnotationView:  readd default column definition for title

to achieve the following:
* use existing lang key that does not match the calculated key
* increase flex size

bwbohl committed:
AnnotationView: improve styling of pos column

* increase width to show full three-digit numbers
* text-align right

bwbohl committed:
AnnotationView: improve styling of pos column

* increase width to show full three-digit numbers
* text-align right

bwbohl committed:
AnnotationView controller: remove unneeded AJAXRequest parameters

The parameters lang and edition are set in window.doAJAXRequest by default

bwbohl committed:
AnnotationView controller: remove unneeded AJAXRequest parameters

The parameters lang and edition are set in window.doAJAXRequest by default

bwbohl committed:
AnnotationView: improve MVC separation of view and controller

bwbohl committed:
AnnotationView: improve MVC separation of view and controller

bwbohl committed:
AnnotationView: do not set column visible if already hidden in default column definition

bwbohl committed:
AnnotationView: do not set column visible if already hidden in default column definition

bwbohl committed:
AnnotationView controller: add debug log

bwbohl committed:
AnnotationView controller: add debug log

bwbohl committed:
AnnotationView: handle undefined emptyFields

bwbohl committed:
AnnotationView: handle undefined emptyFields

bwbohl committed:
AnnotationView:  handle annotation entries/fields with dots in their name

bwbohl committed:
AnnotationView:  handle annotation entries/fields with dots in their name

bwbohl committed:
annotation.scss: apply text-align tight to .annotationList .pos descendants

bwbohl committed:
annotation.scss: apply text-align tight to .annotationList .pos descendants

bwbohl committed:
AnnotationView styling: move todo.css rules to annotation.scss

bwbohl committed:
annotationView styling: move todo.css rules to annotation.scss

bwbohl committed:
AnnotationView: remove duplicate cls property

bwbohl committed:
AnnotationView: remove duplicate cls property

bwbohl commented:
blocked by #101

daniel-jettka reviewed (changes requested):
Built with:

```
export BE_REPO=https://github.com/baz-ga/Edirom-Online-Backend.git FE_REPO=https://github.com/baz-ga/Edirom-Online-Frontend.git BE_BRANCH=ftr/dynamic-annotation-classification FE_BRANCH=ftr/dynamic-annotation-classification && docker compose down --volumes --remove-orphans && docker compose build --no-cache && docker compose up
```

I have some observations:

**(1) Annotation view display**
:x: When I open an annotation (click on annotation marker) in local Klarinettenquintett, the whole annotation list opens and I can't switch to single annotation view. In comparison, [online Klarinettenquintett](https://klarinettenquintett.weber-gesamtausgabe.de/index.html) displays single annotation view by default.


**(2) Annotation view meta info box**
:x: Titles of annotation classes are not shown in single annotation view, and newly added class (I have set up an ediromType category for testing) does not show up here:

<img width="756" height="643" alt="image" src="https://github.com/user-attachments/assets/6e5dd557-7b9e-4ff5-a065-3d36047ee75f" />


**(3) Annotation menu**
:x: Newly added annotation category does not show up in menu:

<img width="938" height="831" alt="image" src="https://github.com/user-attachments/assets/e178b4aa-1917-426b-a9e0-cfcd61889dbe" />


**(4) Annotation list**
:heavy_check_mark: Newly added annotation category shows up in complete annotation list with translation:

<img width="741" height="462" alt="image" src="https://github.com/user-attachments/assets/ed5c9dd9-f937-4139-83c0-d6c1afa4f738" />


**(5) Translation injection**
And one more side-note: Is it possible to inject translations from edition xar into backend? That would be really good for this feature.

bwbohl committed:
Source- and Text-Views: Load taxonomies array instead of categories and priorities

bwbohl committed:
Source- and Text-Views: Load taxonomies array instead of categories and priorities

bwbohl committed:
View – SourceView: use dynamic taxonomies in setAnnotationFilter function

bwbohl committed:
View – SourceView: use dynamic taxonomies in setAnnotationFilter function

bwbohl committed:
View – SourceView: use dynamic taxonomies in annotationFilterChanged function

bwbohl committed:
View – SourceView: use dynamic taxonomies in annotationFilterChanged function

bwbohl committed:
MeasureBasedView annotationFilterChanged: pass taxonomies array to individual viewers

bwbohl committed:
MeasureBasedView annotationFilterChanged: pass taxonomies array to individual viewers

bwbohl committed:
MeasureBasedView – horizontal viewer: process taxonomies array for annotation filter

bwbohl committed:
MeasureBasedView – horizontal viewer: process taxonomies array for annotation filter

bwbohl committed:
MeasureBasedView: fix some indentation and trailing whitespace

bwbohl committed:
MeasureBasedView: fix some indentation and trailing whitespace

bwbohl committed:
MeasureBasedView: add inline documentation

bwbohl committed:
MeasureBasedView: add inline documentation

bwbohl committed:
SourceView: add some inline documentation

bwbohl committed:
SourceView: add some inline documentation

bwbohl committed:
AnnotationView: do not prepend 'view.window.AnnotationView_' for fieldName localisation

bwbohl committed:
AnnotationView: do not prepend 'view.window.AnnotationView_' for fieldName localisation

bwbohl committed:
View PageBasedView – annotationFilter: process taxonomies array

bwbohl committed:
PAgeBasedView – annotationFilter: process taxonomies array

bwbohl committed:
window PageBasedView: cleanup & formatting

bwbohl committed:
View PageBasedView: cleanup & formatting

bwbohl committed:
View PageBasedView: align with View MeasureBasedView

bwbohl committed:
View PageBasedView: align with View MeasureBasedView

bwbohl committed:
View PageBasedView: add some inline documentation

bwbohl committed:
View PAgeBasedView: add some inline documentation

bwbohl committed:
View SourceView: add local-file fallback for taxonomy.label

bwbohl committed:
View SourceView: add local-file fallback for taxonomy.label

bwbohl committed:
View SourceView: add some inline documentation

bwbohl committed:
View SourceView: add some inline documentation

bwbohl committed:
View SourceView: remove trailing whitespaces

bwbohl committed:
View SourceView: remove trailing whitespaces

bwbohl committed:
View TextFacsimileSplitView – setAnnotationFilter: process taxonomies array

bwbohl committed:
View TextFacsimileSplitView – setAnnotationFilter: process taxonomies array

bwbohl committed:
View TextFacsimileSplitView: fix trailing whitespaces and indentation

bwbohl committed:
View TextFacsimileSplitView: fix trailing whitespaces and indentation

bwbohl committed:
View TextView – setAnnotationFilter: process taxonomies array

bwbohl committed:
View TextVies – setAnnotationFilter: process taxonomies array

bwbohl committed:
View TextView: remove trailing whitespace

bwbohl committed:
View TExtView: remove trailing whitespace

bwbohl committed:
AnnotationView: ignore legacyFields

bwbohl committed:
AnnotationView: ignore legacyFields

bwbohl committed:
Application.js: remove call to undefined model.Annotation.updateProxy

bwbohl committed:
Application.js: remove call to undefined model.Annotation.updateProxy

bwbohl committed:
TextFacsimileSplitView: add taxonomyClasses field to annotations store

Add taxonomyClasses field to the annotations data store
configuration in TextFacsimileSplitView to support taxonomy
classification data in annotation responses.

bwbohl committed:
MeasureBasedView: add taxonomyClasses field to annotations store

Add taxonomyClasses field to the annotations data store in
MeasureBasedView to support taxonomy classification data in
annotation responses.

bwbohl committed:
SourceView: add taxonomyClasses field to annotations store

Add taxonomyClasses field to the annotations data store to
support taxonomy classification in source view annotations.
This enables proper handling and display of taxonomy data
returned from the API response.

bwbohl committed:
OpenSeaDragonViewer: support taxonomyClasses in annotation data

Update annotation property retrieval to check for
taxonomyClasses first, falling back to categories for
backward compatibility with existing annotation data
formats.

bwbohl committed:
ImageViewer: support taxonomyClasses in image annotations

Update image viewer to prioritize taxonomyClasses field when
retrieving annotation categories, falling back to the legacy
categories field for backward compatibility with existing
annotations.

bwbohl committed:
feat: pass allTaxonomyIds to annotation filter

Pass allTaxonomyIds parameter through annotation filter
chain to enable proper taxonomy validation. Update filter
logic to skip taxonomies with no matching classes instead
of failing the filter. Refactor to work with annotation
child div IDs rather than annotation objects directly.
Fix typo in comment and improve code formatting.

bwbohl committed:
SourceView controller: merge categories and priorities into taxonomies

Add logic to populate taxonomies array from categories and
priorities when taxonomies is empty. This ensures consistent
data structure for annotation information regardless of the
source data format.

bwbohl committed:
ImageViewer fix: add null check before removing shape element

Add null check before accessing the dom property of
shapeEl to prevent errors when the element does not
exist. This prevents potential null reference exceptions
when removing shape elements from the DOM.

bwbohl committed:
feat: add legacy priority and category columns

Add legacy categories and priority columns to the default columns. These
columns are marked as legacy to indicate they are deprecated fields that
are maintained for backward compatibility with existing annotations.

bwbohl committed:
cleanup: remove unused variable declaration

bwbohl committed:
refactor: addAnnotations in OpenSeaDragonViewer and ImageViewer

Align implementations of addAnnotations methods in
OpenSeadragonViewer and ImageViewer:
- Add comprehensive debug logging
- Consistently rename variables for clarity: outerDiv becomes
  annoIconContainer, innerId becomes annoIconId, and add annoId
  variable.
- Include participant type in annotation icon CSS classes
- Add explanatory comments for code sections and improve code
  organization with better spacing.
- Align logic and code order

bwbohl committed:
fix: hide legacy annotation fields based on preference

Move legacy field hiding logic earlier in the code and apply it
to existing columns. Add hidden property to priority and categories
columns when they are included in the legacyFields array. This
allows users to control visibility of legacy fields through the
annotation_hide_legacy_fields preference while still enabling
re-enablement via the grid header menu.

daniel-jettka commented on app/controller/window/AnnotationView.js:
```diff
@@ -140,5 +133,228 @@ Ext.define('EdiromOnline.controller.window.AnnotationView', {
+                me.createListAndStore(view, data.annotations, me.getStoreFieldsDefinition(view), data.emptyFields);
+
+                // Add the list to the card layout at index 0
+                view.insert(0, view.list);
```
When I comment out lines 178 and 181, it seems to work to load a single annotation instead of the whole list after clicking on a single annotation marker.
Do you remember why you introduced those lines?

bwbohl replied:
I moved those from the view to the controller to better separate the model-view-controller model. Cf. https://github.com/Edirom/Edirom-Online-Frontend/pull/87/changes/068372a13ea92b2cad53baa4845c11e69fe2db6c

bwbohl replied:
could be a race conflict?

bwbohl replied:
Doesn’t happen on my end…

bwbohl replied:
Which data did you use for testing?

daniel-jettka replied:
Thanks for checking. Seems to work now :+1:

bwbohl committed:
fix: prevent late loading annotation view from masking singleView annotation

When a single annotation was opened directly, e.g., from the facsimile
View, a load race condition could replace the annotationView content
with the anniotation list. This commit puts a guard in place to prevent
displaying the list on single annotation calls.

daniel-jettka reviewed (approved):
My previous comment about the list view displayed before single annotation view when an annotation marker is clicked was resolved.

I tested to introduce a new annotation class and it works:

<img width="1121" height="604" alt="Image" src="https://github.com/user-attachments/assets/2abc5cee-d2c4-45d8-905f-8204107a3e0f" />

I have a few suggestions for further enhancement, e.g. displaying (non-legacy) annotation class infos in meta box, but these can be dealt with in a separate feature request.

Nice work, Benni!
